components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/latest/GitpodTerminalService.kt

```diff
@@ -25,43 +28,32 @@ import java.util.concurrent.CompletableFuture
 import java.util.concurrent.TimeUnit
 
 @Suppress("UnstableApiUsage")
-class GitpodTerminalService(session: ClientProjectSession) : Disposable {
+class GitpodTerminalService(private val session: ClientProjectSession) {
     private companion object {
-        /** Indicates if this service is already running, because we shouldn't run it more than once. */
-        var isRunning = false
+        var hasCreatedTerminals = false
+        val forwardedPortsList: MutableSet<Int> = mutableSetOf()
     }
 
-    private val lifetime = Lifetime.Eternal.createNested()
     private val terminalView = TerminalView.getInstance(session.project)
     private val backendTerminalManager = BackendTerminalManager.getInstance(session.project)
+    private val portForwardingManager = PortForwardingManager.getInstance(session.project)
     private val terminalServiceFutureStub = TerminalServiceGrpc.newFutureStub(GitpodManager.supervisorChannel)
     private val statusServiceStub = StatusServiceGrpc.newStub(GitpodManager.supervisorChannel)
 
-    override fun dispose() {
-        lifetime.terminate()
-    }
-
-    init {
-        run()
-    }
+    init { start() }
 
-    private fun run() {
-        if (application.isHeadlessEnvironment || isRunning) return
+    private fun start() {
+        if (application.isHeadlessEnvironment || hasCreatedTerminals) return
 
-        isRunning = true
-
-        val task = application.executeOnPooledThread {
+        application.executeOnPooledThread {
             val terminals = getSupervisorTerminalsList()
             val tasks = getSupervisorTasksList()
 
             application.invokeLater {
                 createTerminalsAttachedToTasks(terminals, tasks)
+                hasCreatedTerminals = true
```

Comment:
then you need to move it back after 
`if (application.isHeadlessEnvironment || hasCreatedTerminals) return`

otherwise it can on the client reopening before createTerminalsAttachedToTasks executed

Reply:
💡  True!
Renamed it to `hasStarted` and moved to the next line after that conditional.